docs/extended-collection-types.md

```diff
@@ -0,0 +1,169 @@
+
+Example usage:
+```
+var/odlist/foo = odlist("a","b","c")
+
+foo += "d" // Valid modification, just like a BYOND list
+
+foo["c"] = 5 // Compile-time error, ODLists cannot be associative
+
+world.log << foo.Join() // Prints "abcd"
+```
+
+### Queue
+Similar to lists, except elements are first-in first-out. Most of the list API will remain, except for inserting and removing elements. These will use `queue.enqueue(element)` and `queue.dequeue()`.
+
+Example usage:
+```
+var/queue/foo = queue("a","b","c")
+
+foo += "d" // Compile-time error, must use foo.enqueue("d")
+
+world.log << foo[2] // Prints "b", indexing is supported
+
+world.log << foo.dequeue() // Removes "a" from `foo` and prints "a"
+```
+
+A `queue.to_array()` helper will also be provided, and returns the queue's elements as a shallow-copied `array()`.
+
+### Stack 
+Similar to queues, except elements are last-in first-out. Most of the list API will remain, except for inserting and removing elements. These will use `stack.push(element)` and `stack.pop()`.
+
+Example usage:
+```
+var/stack/foo = stack("a","b","c")
+
+foo += "d" // Compile-time error, must use foo.push("d")
+
+world.log << foo[2] // Prints "b", indexing is supported
+
+foo.push("d")
+world.log << foo.pop() // Removes "d" from `foo` and prints "d"
+```
+
+A `stack.to_array()` helper will also be provided, and returns the stack's elements as a shallow-copied `array()`.
+
+## Constant Collections
+
+All OpenDream collections can now be declared as constant, which prevents any future modifications to the collection contents and will enable future internal optimizations.
+
+Example usage:
+```
+var/const/array/foo = array("a","b","c")
+foo[1] = "z" // Compile-time error, cannot modify constant collection
+```
+
+Note that `var/const/list/L` is not valid in BYOND and usages of constant collections will require a wrapper macro.
+
+## Collection Capacity API
+
+Collections (other than the existing BYOND `list()` and the fixed-length `array()` types) will expose the C# `Capacity` property and `EnsureCapacity()` functionality. These operate similarly to `list.len`, except capacity refers to the number of elements the collection's internal data structure can hold without allocating more memory. Ergo a list could have 3 elements and a capacity of 5, and no memory allocation will occur until a 6th element is inserted.
+
+Accessing `list.capacity` (and similar for other collection types) will return the currect value of the C# `List.Capacity`. Setting it will set it in C#, which may runtime if the new capacity is smaller than the current length of the collection.
+
+Calling `list.ensure_capacity(size)` (and similar for other collection types) will call the C# method `List.EnsureCapacity(size)`. This method ensures that the capacity of this collection is at least the specified capacity. If the current capacity is less than capacity, it is increased to at least the specified capacity.
```

Comment:
```suggestion
Accessing `collection.capacity` (and similar for other collection types) will return the currect value of the C# `Capacity` property. Setting it will set it in C#, which may runtime if the new capacity is smaller than the current length of the collection.

Calling `collection.ensure_capacity(size)` will call the C# method `EnsureCapacity(size)`. This method ensures that the capacity of this collection is at least the specified capacity. If the current capacity is less than capacity, it is increased to at least the specified capacity.
```